spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,84 +120,138 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
-					if (StringUtils.hasText(streamListener.value())) {
-						for (int i = 0; i < parameterTypes.length; i++) {
-							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, i);
-							Assert.isTrue(methodParameter.getParameterAnnotation(Input.class) == null &&
-											methodParameter.getParameterAnnotation(Output.class) == null,
-									"A message handling @StreamListener method cannot have parameters annotated " +
-											"with @Input or @Output");
-						}
-						if (!method.getReturnType().equals(Void.TYPE)) {
-							Assert.isTrue(method.getAnnotation(Output.class) == null,
-									"A message handling @StreamListener method cannot be annotated with @Output");
+					if (!method.getReturnType().equals(Void.TYPE)) {
+						if (!StringUtils.hasText(methodAnnotatedOutboundName)) {
+							if (outputAnnotationCount == 0) {
+								throw new IllegalArgumentException(RETURN_TYPE_NO_OUTBOUND_SPECIFIED);
+							}
+							Assert.isTrue((outputAnnotationCount == 1), RETURN_TYPE_MULTIPLE_OUTBOUND_SPECIFIED);
 						}
-						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
+					}
+					if (isDeclarative && (!StringUtils.hasText(methodAnnotatedInboundName) || parameterTypes.length == 1)) {
+						invokeSetupMethodOnListenedChannel(method, bean, methodAnnotatedInboundName, methodAnnotatedOutboundName);
 					}
 					else {
-						for (int i = 0; i < parameterTypes.length; i++) {
-							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, i);
-							Assert.isTrue(methodParameter.getParameterAnnotation(Input.class) != null ^
-											methodParameter.getParameterAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must have its parameters annotated" +
-											"with @Input or @Output, but not with both.");
-						}
-						if (!method.getReturnType().equals(Void.TYPE)) {
-							Assert.isTrue(method.getAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must be annotated with @Output");
-						}
-						invokeSetupMethodOnListenedChannel(method, bean);
+						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
 					}
 				}
 			}
 		});
 		return bean;
 	}
 
+	private boolean isDeclarativeStreamListenerMethod(Method method, String methodAnnotatedInboundName, String methodAnnotatedOutboundName) {
+		int methodArgumentsLength = method.getParameterTypes().length;
+		for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Input.class)) {
+				String inboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Input.class));
+				Assert.isTrue(StringUtils.hasText(inboundName), INVALID_INBOUND_NAME);
+				return isDeclarativeMethodParameter(getBindableBean(inboundName), methodParameter);
```

Comment:
A parameter annotated with `@Input` must be declarative so here we could just throw an error if it isn't.